From: Ray Strode <rstrode@redhat.com>
Date: Wed, 7 Feb 2024 11:47:32 -0500
Subject: gdm: Add "Login Failed" notification

One step toward implementing Web Login is adding the ability to
show "Login Failed" as a notification banner.

This is because with Web Login, the auth prompt is occluded so can't
be used to show an error message itself.

This commit adds the necessary code to show the Login Failed
notification, but doesn't actually call that code yet.

It will be used later when Web Login support is added.
---
 .../gnome-shell-sass/widgets/_login-lock.scss      | 13 ++++++++++++
 js/gdm/authPrompt.js                               | 23 ++++++++++++++++++++++
 2 files changed, 36 insertions(+)

diff --git a/data/theme/gnome-shell-sass/widgets/_login-lock.scss b/data/theme/gnome-shell-sass/widgets/_login-lock.scss
index f9183f1..c7411d0 100644
--- a/data/theme/gnome-shell-sass/widgets/_login-lock.scss
+++ b/data/theme/gnome-shell-sass/widgets/_login-lock.scss
@@ -393,6 +393,19 @@ $_gdm_dialog_width: 25em;
   }
 }
 
+// Login Failed banner
+.auth-notification-banner {
+  .message-header {
+    width: 0px;
+    height: 0px;
+
+    .message-close-button {
+      border: none;
+      background-color: rgba(0, 0, 0, 0.0);
+    }
+  }
+}
+
 // QR Code
 .qr-code {
   background: black;
diff --git a/js/gdm/authPrompt.js b/js/gdm/authPrompt.js
index eb28eb0..3c8e1f6 100644
--- a/js/gdm/authPrompt.js
+++ b/js/gdm/authPrompt.js
@@ -10,6 +10,7 @@ import * as Animation from '../ui/animation.js';
 import * as AuthList from './authList.js';
 import * as Batch from './batch.js';
 import * as GdmUtil from './util.js';
+import * as MessageTray from '../ui/messageTray.js';
 import * as Params from '../misc/params.js';
 import * as ShellEntry from '../ui/shellEntry.js';
 import * as UserWidget from '../ui/userWidget.js';
@@ -650,6 +651,28 @@ export const AuthPrompt = GObject.registerClass({
         this._userVerifier.setForegroundMechanism(mechanism);
     }
 
+    showLoginFailedNotification() {
+        const source = new MessageTray.getSystemSource();
+
+        this._loginFailedNotification = new MessageTray.Notification({
+            source,
+            iconName: 'dialog-password-symbolic',
+            title: _('Login Failed'),
+            body: _('Please try again'),
+            isTransient: true,
+        });
+
+        source.addNotification(this._loginFailedNotification);
+    }
+
+    hideLoginFailedNotification() {
+        if (!this._loginFailedNotification)
+            return;
+
+        this._loginFailedNotification.destroy();
+        this._loginFailedNotification = null;
+    }
+
     reset(params) {
         let {beginRequestType, queryingService} = Params.parse(params, {
             beginRequestType: null,
